model/pipeline/src/main/proto/beam_runner_api.proto

```diff
@@ -698,10 +700,10 @@ message StandardCoders {
 // TODO: consider inlining field on PCollection
 message WindowingStrategy {
 
-  // (Required) The SdkFunctionSpec of the UDF that assigns windows,
+  // (Required) The FunctionSpec of the UDF that assigns windows,
   // merges windows, and shifts timestamps before they are
   // combined according to the OutputTime.
-  SdkFunctionSpec window_fn = 1;
+  FunctionSpec window_fn = 1;
```

Comment:
All other uses of SdkFunctionSpec are part of a transform but finding the envrionment for the windowing strategy that the window_fn is part of when looking at a PCollection is difficult since one needs to find the environment that the upstream assign windows transform is part of. It is possible but annoying. Also this is different then coders which don't have an environment since both the upstream and downstream transforms need to understand the encoding and runners have the length prefixing technique to convert coders from unknown ones to known ones.

For now I'm for not adding an environment id here and forcing that graph traversal for the assign windows transform to find the environment and in the future to truly support custom window fns its likely we will have to come up with techniques like the coder length prefixing to make it so that they can be treated opaquely. (note that rest of the change looks good)

@chamikaramj @robertwb what do you think?

Reply:
That makes sense and I'm actually running into this issue while updating SDKs. We probably have to add environment_id to WindowingStrategy since it's not directly attached to PTransform.

Reply:
The windowing strategy cannot be coerced into an environment-agnostic one the same way a coder can be, as it involves actually executing user code as oppose to specifying constraints on its behavior. (In some sense, it is in a very real sense a cross-language UDF.) 

Having to traverse the graph is ugly, but I suppose possible for now.

Reply:
After hearing about some of the difficulties that Cham is running into. I would go either way on whether we add an environment id here or remove it completely.

Reply:
Added an environment_id to WindowingStrategy for now (retaining the current behavior). This can be further simplified in the future if needed.